src/shell/io.c

```diff
@@ -101,52 +100,58 @@ static int shell_io_write (struct shell_io *io,
                            int rank,
                            const char *stream,
                            const char *data,
-                           int len)
+                           int len,
+                           bool eof)
 {
-    flux_future_t *f;
+    flux_future_t *f = NULL;
+    json_t *o = NULL;
 
-    if (!(f = shell_svc_pack (io->shell->svc,
-                             "write",
-                             0,
-                             0,
-                             "{s:i s:s s:i s:s}",
-                             "rank", rank,
-                             "stream", stream,
-                             "len", len,
-                             "data", data)))
-        return -1;
-    if (flux_future_then (f, -1, shell_io_write_completion, io) < 0) {
-        flux_future_destroy (f);
+    if (!(o = ioencode (stream, rank, data, len, eof))) {
+        log_err ("ioencode");
         return -1;
     }
+
+    if (!(f = shell_svc_pack (io->shell->svc,
+                              "write",
+                              0,
+                              0,
+                              "{s:O}",
+                              "io", o)))
```

Comment:
ahhh, good catch.  IIRC `flux_rpc()` requires payloads to be json objects.  So I instinctively always create one.  But in this case, the ioencode creates a json object, so it can go straight in.